benches/distribution_statistic.rs

```diff
@@ -0,0 +1,73 @@
+use criterion::{criterion_group, criterion_main, BatchSize, Criterion};
+
+use rand::distributions::Distribution;
+use rand::distributions::Uniform;
+use rand::rngs::SmallRng;
+use rand::seq::SliceRandom;
+use rand::SeedableRng;
+
+use vector::event::metric::Sample;
+use vector::sinks::util::statistic::DistributionStatistic;
+
+fn generate_samples(mut size: u32, max_bin_count: u32) -> Vec<Sample> {
+    // generate random samples, but we also want to use
+    // the same samples set on each run.
+    let mut rng = SmallRng::seed_from_u64(1234);
+    let range = Uniform::from(1..=max_bin_count);
+    let mut value = 1.0;
+    let mut samples = Vec::new();
+    while size > 0 {
+        let bin_count = u32::min(range.sample(&mut rng), size);
+        samples.push(Sample {
+            value,
+            rate: bin_count,
+        });
+        size -= bin_count;
+        value += 1.0;
+    }
+    samples.shuffle(&mut rng);
+    samples
+}
+
+fn bench_statistic(c: &mut Criterion) {
+    let mut group = c.benchmark_group("distribution_statistic");
+
+    let sizes = [5, 10, 50, 100, 200, 500, 1000];
+    for &size in &sizes {
+        group.bench_function(format!("small-bin-{}", size), |b| {
+            let samples = generate_samples(size, 3);
+
+            b.iter_batched(
+                || samples.clone(),
```

Comment:
By using the same set of samples each time, this may be benchmarking the branch predictor in the CPU more than it is the underlying code.

I rewrote this using `thread_rng` above instead of a fixed seed and generating a new set of samples each batch instead of cloning. It did increase the noise but only up to about 2% at the worst. There are a *lot* of iterations run, and that strongly averages out any differences introduced by differing batches.

Reply:
I've moved `generating_samples` to `iter_batched`'s setup and increased noise threshold to 0.1.